### PR TITLE
Hector.vote snapshot proposal

### PR DESCRIPTION
**Pull request by fstar1129**
## Description
Voting on a Snapshot Proposal, this creates a Vote object/receipt that needs to be saved to IPFS (in a fetch-able way). (PR8)
UI component for submitting a vote that exists on the Snapshot Proposal Page. Only interact-able if you are validated by the strategy.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [ ] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [ ] yes, and they are not tested: [enter the % coverage here]
- [ ] yes, but I did not run tests
- [ ] no

### CODE REVIEW

File: client/scripts/helpers/snapshot_utils/networks.json [hunk 1]
```diff
@@ -0,0 +1,300 @@
+{
```
Comment (zakhap) [inline client/scripts/helpers/snapshot_utils/networks.json:1]: why is this file necessary? can we just fetch it to the client if necessary?

File: client/scripts/helpers/snapshot_utils/snapshot_client.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import Snapshot from '@snapshot-labs/snapshot.js';
+
+const hubUrl = process.env.SNAPSHOT_APP_HUB_URL || 'https://testnet.snapshot.org';
+const snapshotClient = new Snapshot.Client(hubUrl);
+
+export default snapshotClient;
```
Comment (zakhap) [inline client/scripts/helpers/snapshot_utils/snapshot_client.ts:6]: This should be added to the snapshot controller so that the client can just be generated once as part of a class

Comment (zakhap) [reply client/scripts/helpers/snapshot_utils/snapshot_client.ts:6]: rather that it being instantiated each time it's imported

File: client/scripts/helpers/snapshot_utils/snapshot_utils.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import numeral from 'numeral';
+import Snapshot from '@snapshot-labs/snapshot.js';
+
+import client from 'helpers/snapshot_utils/snapshot_client';
+
+const gateway = process.env.SNAPSHOT_IPFS_GATEWAY || gateways[0];
+
+export function jsonParse(input, fallback?) {
+	if (typeof input !== 'string') {
+		return fallback || {};
```
Comment (zakhap) [inline client/scripts/helpers/snapshot_utils/snapshot_utils.ts:12]: fix linting for this whole file

File: client/scripts/views/modals/confirm_snapshot_vote_modal.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	selectedChoice,
+	totalScore,
+	scores,
+	snapshot
+}, {
+  error: any,
+  saving: boolean,
+}> = {
+  view: (vnode) => {
+		const author = app.user.activeAccount;
```
Comment (zakhap) [inline client/scripts/views/modals/confirm_snapshot_vote_modal.ts:31]: fix linting on this file

File: client/scripts/views/modals/confirm_snapshot_vote_modal.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+									type: 'vote',
+									payload: {
+										proposal: id,
+										choice: selectedChoice + 1,
+										metadata: {}
+									}
+								})
+							};
+							const msgBuffer = bufferToHex(new Buffer(msg.msg, 'utf8'));
+  						msg.sig = await (window as any).ethereum.request({method: 'personal_sign', params: [msgBuffer, author.address]});
```
Comment (zakhap) [inline client/scripts/views/modals/confirm_snapshot_vote_modal.ts:95]: also please don't use window.ethereum for message signing requests. look elsewhere in our repo for how submit transactions and get signatures. was this just how snapshots docs did it?

File: client/scripts/views/modals/tx_signing_modal.ts [hunk 6]
```diff
@@ -430,14 +430,14 @@ const TXSigningModalStates: {
 
       vnode.state.timer = 0;
       // TODO: set a timeout? We currently have no failure case due to how event handling works.
-      vnode.state.timerHandle = setInterval(() => {
+      vnode.state.timerHandle = global.setInterval(() => {
```
Comment (zakhap) [inline client/scripts/views/modals/tx_signing_modal.ts:433]: what is this setTimeout and why does it have to be global?

File: client/scripts/views/modals/tx_signing_modal.ts [hunk 7]
```diff
@@ -430,14 +430,14 @@ const TXSigningModalStates: {
...
         vnode.state.timer++;
         m.redraw();
       }, 1000);
       // for edgeware mainnet, timeout after 10 sec
       // TODO: remove this after the runtime upgrade to Substrate 2.0 rc3+
       if (app.chain?.meta?.chain?.id === 'edgeware') {
-        vnode.state.timeoutHandle = setTimeout(() => {
+        vnode.state.timeoutHandle = global.setTimeout(() => {
```
Comment (zakhap) [inline client/scripts/views/modals/tx_signing_modal.ts:440]: why did this need to be changed for `edgeware`?

File: client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts [hunk 8]
```diff
@@ -105,14 +92,14 @@ const newThread = async (
...
-  form.snapshot = await getBlockNumber(getProvider(space.network));
+  form.snapshot = await snapshotJs.utils.getBlockNumber(snapshotJs.utils.getProvider(space.network));
   form.metadata.network = space.network;
   form.metadata.strategies = space.strategies;
 
   const msg: any = {
     address: author.address,
     msg: JSON.stringify({
-      version,
+      version: '0.1.3',
```
Comment (zakhap) [inline client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts:102]: why do we need to specify the version here? can we fetch the appropriate version from elsewhere?

File: client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts [hunk 9]
```diff
@@ -175,34 +162,30 @@ export const NewProposalForm: m.Component<{snapshotId: string}, {
     };
 
     snapshotClient.getSpaces().then(response => {
-      let spaces: any = fromEntries(
+      let spaces: any = Object.fromEntries(
```
Comment (zakhap) [inline client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts:165]: there are better ways to do this

Comment (zakhap) [reply client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts:165]: I'm not sure i understand why this is done this way

File: client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts [hunk 10]
```diff
@@ -175,34 +162,30 @@ export const NewProposalForm: m.Component<{snapshotId: string}, {
...
-      //   m.redraw();
-      // });
+
+      snapshotJs.utils.getScores(
+        space.key,
+        space.strategies,
+        space.network,
+        snapshotJs.utils.getProvider(space.network),
+        [app.user.activeAccount.address]
+      ).then(response => {
```
Comment (zakhap) [inline client/scripts/views/pages/new_snapshot_proposal/new_proposal_form.ts:179]: same here as above.

File: client/scripts/views/pages/view_snapshot_proposal/index.ts [hunk 11]
```diff
@@ -115,6 +117,11 @@ const VoteView: m.Component<{ votes: Vote[] }> = {
 };
 
 const VoteAction: m.Component<{
+  space: any,
+	proposal: any,
```
Comment (zakhap) [inline client/scripts/views/pages/view_snapshot_proposal/index.ts:121]: lint

Comment (zakhap) [reply client/scripts/views/pages/view_snapshot_proposal/index.ts:121]: this whole file too

File: client/scripts/views/pages/view_snapshot_proposal/index.ts [hunk 12]
```diff
@@ -179,38 +209,64 @@ const ViewProposalPage: m.Component<{
...
-                timestamp: '',
-              };
-              vote.voterAddress = key,
-              vote.timestamp = response[key].msg.timestamp;
-              vote.choice = response[key].msg.payload.choice === 1 ? 'yes' : 'no';
-              votes.push(vote);
-            }
-            vnode.state.votes = votes;
-            m.redraw();
+      snapshotClient.getSpaces().then(response => {
```
Comment (zakhap) [inline client/scripts/views/pages/view_snapshot_proposal/index.ts:227]: why do we do this?

Comment (zakhap) [reply client/scripts/views/pages/view_snapshot_proposal/index.ts:227]: This feels like something that should happen at snapshot controller initialization, stored on the class, and then utilized top down, rather than each time a page/view/component loads within snapshot.

File: client/styles/modals/confirm_snapshot_vote_modal.scss [hunk 13]
```diff
@@ -0,0 +1,40 @@
+@import 'client/styles/shared';
+
+.ConfirmSnapshotVoteModal {
```
Comment (zakhap) [inline client/styles/modals/confirm_snapshot_vote_modal.scss:3]: lint file

File: server/socket/index.ts [hunk 14]
```diff
@@ -64,10 +64,10 @@ export default function (
 
           // reset liveness timers
           if (ws.aliveTimer) clearTimeout(ws.aliveTimer);
-          ws.aliveTimer = setTimeout(() => { ws.isAlive = false; }, ALIVE_TIMEOUT);
+          ws.aliveTimer = global.setTimeout(() => { ws.isAlive = false; }, ALIVE_TIMEOUT);
 
           if (ws.expirationTimer) clearTimeout(ws.expirationTimer);
-          ws.expirationTimer = setTimeout(() => {
+          ws.expirationTimer = global.setTimeout(() => {
```
Comment (zakhap) [inline server/socket/index.ts:70]: why are these globals updated?

File: server/util/cacheJobRunner.ts [hunk 15]
```diff
@@ -19,7 +19,7 @@ export default abstract class JobRunner<CacheT> {
 
   public start(...args) {
     if (this._jobTimeS > 0) {
-      this._timeoutHandle = setInterval(() => this.run(), this._jobTimeS * 1000);
+      this._timeoutHandle = global.setInterval(() => this.run(), this._jobTimeS * 1000);
```
Comment (zakhap) [inline server/util/cacheJobRunner.ts:22]: why global?

**Review by zakhap — CHANGES_REQUESTED** [review]
There are a bunch of things here that should be abstracted into the controller level so they don't get instantiated over and over. If you have any questions about what I mean anywhere, please don't hesitate to ask.

File: client/scripts/helpers/snapshot_utils/gateways.json [hunk 16]
```diff
@@ -0,0 +1,9 @@
+[
```
[the remark this answers is not in the document]
Comment (fstar1129) [reply client/scripts/helpers/snapshot_utils/gateways.json:1]: It is IPFS gateway to get proposal data from IPFS. They did not export it. So I should make a new file for it.

File: client/scripts/helpers/snapshot_utils/networks.json (hunk 1 above)
Comment (fstar1129) [reply client/scripts/helpers/snapshot_utils/networks.json:1]: This file is used to get scan explorer url. They did not export this file. So can not fetch it from the client.

File: client/scripts/helpers/snapshot_utils/snapshot_client.ts (hunk 2 above)
Comment (fstar1129) [reply client/scripts/helpers/snapshot_utils/snapshot_client.ts:6]: Got it

File: client/scripts/helpers/snapshot_utils/snapshot_utils.ts (hunk 3 above)
Comment (fstar1129) [reply client/scripts/helpers/snapshot_utils/snapshot_utils.ts]: Got it

File: client/scripts/views/modals/confirm_snapshot_vote_modal.ts (hunk 4 above)
Comment (fstar1129) [reply client/scripts/views/modals/confirm_snapshot_vote_modal.ts:31]: Got it

File: client/scripts/views/pages/view_snapshot_proposal/index.ts (hunk 11 above)
Comment (fstar1129) [reply client/scripts/views/pages/view_snapshot_proposal/index.ts:120]: Got it

File: client/scripts/views/pages/view_snapshot_proposal/index.ts (hunk 12 above)
Comment (fstar1129) [reply client/scripts/views/pages/view_snapshot_proposal/index.ts]: Correct!

File: client/styles/modals/confirm_snapshot_vote_modal.scss (hunk 13 above)
Comment (fstar1129) [reply client/styles/modals/confirm_snapshot_vote_modal.scss:3]: Got it

File: server/socket/index.ts (hunk 14 above)
Comment (fstar1129) [reply server/socket/index.ts:70]: I updated tsconfig adding esnext in lib to use a functionality which is not in es6. Then setTimeout made a type error. So I added it global then it is fixed.

File: server/util/cacheJobRunner.ts (hunk 15 above)
Comment (fstar1129) [reply server/util/cacheJobRunner.ts:22]: I wrote a reason above

**Comment by gdjohnson** [comment]
Struggling to get this to compile—
```
ERROR in /Users/suspendedreason/commonwealth/client/scripts/helpers/snapshot_utils/snapshot_utils.ts
./client/scripts/helpers/snapshot_utils/snapshot_utils.ts
[tsl] ERROR in /Users/suspendedreason/commonwealth/client/scripts/helpers/snapshot_utils/snapshot_utils.ts(56,27)
      TS2339: Property 'getVotes' does not exist on type 'Client'.

ERROR in /Users/suspendedreason/commonwealth/node_modules/@snapshot-labs/snapshot.js/node_modules/ajv/dist/types/json-schema.d.ts
[tsl] ERROR in /Users/suspendedreason/commonwealth/node_modules/@snapshot-labs/snapshot.js/node_modules/ajv/dist/types/json-schema.d.ts(1,92)
      TS1005: '?' expected.

ERROR in /Users/suspendedreason/commonwealth/node_modules/@snapshot-labs/snapshot.js/node_modules/ajv/dist/types/json-schema.d.ts
[tsl] ERROR in /Users/suspendedreason/commonwealth/node_modules/@snapshot-labs/snapshot.js/node_modules/ajv/dist/types/json-schema.d.ts(1,151)
      TS1005: ';' expected.
```

Any suggestions @fstar1129 ?

**Comment by zakhap** [comment]
@fstar1129 can you also clarify why we needed `esnext`, what features are needed beyond what's in es6?

**Comment by fstar1129** [comment]
Snapshot.js is updated. Let me fix it.

**Comment by fstar1129** [comment]
es6 does not support Object.fromEntries.
https://www.typescriptlang.org/tsconfig#lib

**Comment by zakhap** [comment]
> es6 does not support Object.fromEntries.

Why do we need this? We can just convert these things to more vanilla js

**Comment by zakhap** [comment]
@fstar1129 This still doesn't compile

**Comment by zakhap** [comment]
@fstar1129 this doesn't compile still

**Comment by fstar1129** [comment]
> @fstar1129 this doesn't compile still

I am working on it.

**Comment by gdjohnson** [comment]
When the `ConfirmSnapshotVote` modal gets created on voting, it passes data incl. a `snapshot` property that's given `vnode.attrs.proposal.msg.payload.snapshot`—but the Sushi proposals I'm playing with don't have a `msg` property on proposals, so the voting fails silently.

I'm adding try/catch structure with a `notifyError` call, so that users know when their voting is failing—but can the root problem get addressed?
